vision_datasets/dataset/vision_dataset.py

```diff
@@ -0,0 +1,176 @@
+        # Convert absolute coordinates to relative coordinates.
+        # Example: for image with size (200, 200), (1, 100, 100, 200, 200) => (1, 0.5, 0.5, 1.0, 1.0)
+        if dataset_info.type == DatasetTypes.MULTITASK:
+            return {task_name: VisionDataset._convert_box_to_relative(task_target, img_w, img_h, dataset_info.sub_task_infos[task_name]) for task_name, task_target in target.items()}
+
+        if dataset_info.type == DatasetTypes.IMAGE_OBJECT_DETECTION:
+            relative_target = copy.deepcopy(target)
+            for t in relative_target:
+                label = t.label_data
+                t.label_data = [label[0], label[1] / img_w, label[2] / img_h, label[3] / img_w, label[4] / img_h]
+            return relative_target
+
+        return target
+
+
+class LocalFolderCacheDecorator(BaseDataset):
+    """
+    Decorate a dataset by caching data in a local folder, in local_cache_params['dir'].
+
+    """
+
+    def __init__(self, dataset: BaseDataset, local_cache_params: dict):
+        """
+        Args:
+            dataset: dataset that requires cache
+            local_cache_params(dict): params controlling local cache for image access:
+                'dir': local dir for caching crops, it will be auto-created if not exist
+                [optional] 'n_copies': default being 1. if n_copies is greater than 1, then multiple copies will be cached and dataset will be n_copies times bigger
+        """
+
+        assert dataset is not None
+        assert local_cache_params
+        assert local_cache_params.get('dir')
+        local_cache_params['n_copies'] = local_cache_params.get('n_copies', 1)
+        assert local_cache_params['n_copies'] >= 1, 'n_copies must be equal or greater than 1.'
+
+        super().__init__(dataset.dataset_info)
+
+        self._dataset = dataset
+        self._local_cache_params = local_cache_params
+        if not os.path.exists(self._local_cache_params['dir']):
+            os.makedirs(self._local_cache_params['dir'])
+
+        self._annotations = {}
+        self._paths = {}
+
+    @property
+    def labels(self):
+        return self._dataset.labels
+
+    def __len__(self):
+        return len(self._dataset) * self._local_cache_params['n_copies']
+
+    def _get_single_item(self, index):
+        annotations = self._annotations.get(index)
+        if annotations:
+            return Image.open(self._paths[index]), annotations, str(index)
+
+        idx_in_epoch = index % len(self._dataset)
+        img, annotations, _ = self._dataset[idx_in_epoch]
+        local_img_path = self._construct_local_image_path(index, img.format)
+        self._save_image_matching_quality(img, local_img_path)
+        self._annotations[index] = annotations
+        self._paths[index] = local_img_path
+
+        return img, annotations, str(index)
+
+    def _construct_local_image_path(self, img_idx, img_format):
+        return pathlib.Path(self._local_cache_params['dir']) / f'{img_idx}.{img_format}'
```

Comment:
why don't we make pathlib.Path() in the constructor?

Reply:
done